src/VirtoCommerce.Platform.Web/Controllers/Api/ModulesController.cs

```diff
@@ -90,11 +94,50 @@ public ActionResult<ModuleDescriptor[]> GetModules()
         {
             EnsureModulesCatalogInitialized();
 
-            var retVal = _externalModuleCatalog.Modules.OfType<ManifestModuleInfo>().OrderBy(x => x.Id).ThenBy(x => x.Version)
-                                               .Select(x => new ModuleDescriptor(x))
-                                               .ToArray();
+            var allModules = _externalModuleCatalog.Modules
+                .OfType<ManifestModuleInfo>()
+                .OrderBy(x => x.Id)
+                .ThenBy(x => x.Version)
+                .Select(x => new ModuleDescriptor(x))
+                .ToList();
 
-            return Ok(retVal);
+            _localModuleCatalog.Initialize();
+            var localModules = _localModuleCatalog.Modules.OfType<ManifestModuleInfo>().ToDictionary(x => x.Id);
+
+            foreach (var module in allModules.Where(x => !string.IsNullOrEmpty(x.IconUrl)))
+            {
+                if (!localModules.TryGetValue(module.Id, out var localModule) ||
+                    !IconFileExists(localModule))
+                {
+                    module.IconUrl = null;
+                }
+            }
+
+            return Ok(allModules);
+        }
+
+        private static bool IconFileExists(ManifestModuleInfo module)
+        {
+            // PathString should start from "/"
+            var moduleIconUrl = module.IconUrl;
+            if (!moduleIconUrl.StartsWith('/'))
+            {
+                moduleIconUrl = "/" + moduleIconUrl;
+            }
+
+            var basePath = new PathString($"/modules/$({module.Id})");
+            var iconUrlPath = new PathString(moduleIconUrl);
+
+            if (iconUrlPath.StartsWithSegments(basePath, out var subPath) && !string.IsNullOrEmpty(subPath.Value))
```

Comment:
For some modules this doesn't work because icon URL contains invalid module ID.
For example, in VirtoCommerce.**Assets** module the icon URL is Modules/$(VirtoCommerce.**AssetsModule**)/Content/logo.png